### ISSUE TITLE
🐛 Linker error on FreeBSD

### ISSUE THREAD

**Issue by yurivict** [cda-tum/mqt-core#503]
### Environment information

Version: 2.0.2
clang-16
FreeBSD 13.2

### Description

Build fails:
```
ld: error: undefined symbol: testing::internal2::PrintBytesInObjectTo(unsigned char const*, unsigned long, std::__1::basic_ostream<char, std::__1::char_traits<char>>*)
>>> referenced by gtest.h:8016 (/usr/local/include/gtest/gtest.h:8016)
>>>               test/CMakeFiles/mqt-core-test-zx.dir/zx/test_zx.cpp.o:(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> testing::PrintToString<zx::EdgeType>(zx::EdgeType const&))
>>> referenced by gtest.h:8016 (/usr/local/include/gtest/gtest.h:8016)
>>>               test/CMakeFiles/mqt-core-test-zx.dir/zx/test_zx.cpp.o:(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> testing::PrintToString<zx::VertexType>(zx::VertexType const&))
c++: error: linker command failed with exit code 1 (use -v to see invocation)
```

### How to Reproduce

Regular build.

**Comment by burgholzer** [comment]
Hey 👋🏼 
Thanks for raising this issue. I am currently in the process of setting up a FreeBSD VM to test this myself.
Are you sure that this is a "regular build"?
For some reason, the googletest version that is being referenced here
> referenced by gtest.h:8016 (/usr/local/include/gtest/gtest.h:8016)

is a locally installed one. The `mqt-core` project does ship a version of googletest as a submodule though (in `extern/googletest`. And a regular build should link to that submodule.
A quick search for the error reveals https://github.com/google/googletest/issues/3816 where it is also stated that this can be due to incompatible versions of googletest.

I'd like to understand what makes FreeBSD so special that this fails under that system but passes on every other system we test (Ubuntu 22.04, macOS 12, Windows 2022). Maybe you can provide some context.

**Comment by yurivict** [comment]
Hi,

> Are you sure that this is a "regular build"?

The build is done in the FreeBSD ports' framework using cmake in a regular way, like all other cmake-based ports build.

> [...] the googletest version that is being referenced here [...] is a locally installed one.

This means that the pre-installed googletest is used when it is installed instead of the bundled one.
This has to do with compiler paths.

> Maybe you can provide some context.

FreeBSD has its own ports/packages infrastructure. Also all local packages are installed into /usr/local, not into /usr like in Linux. Maybe this is what makes it "special".

**Comment by burgholzer** [comment]
Thanks for the reply. I'll just have to try myself whenever the system is done setting up because I still do not quite understand how the pre-installed `googletest` can even be used. 
We explicitly do not use something like `find_packages` or `FetchContent` at the moment and do a simple `add_subdirectory` for the included submodule.
We do guard that inclusion via 
> if(NOT TARGET gtest OR NOT TARGET gmock)

in order to avoid redefining a target, but I fail to see how the `gtest` target were defined already when invoking CMake.

Also, according to https://www.freshports.org/devel/googletest/ FreeBSD 13 is using `googletest` version 1.14, which is pretty close to what we are using (current HEAD).
Maybe all of this would be solved if we didn't rely on submodules and instead used `FetchContent` right away.

**Comment by yurivict** [comment]
pre-installed googletest is used likely due to the system path being the first.

Please don't use fetchcontent because fetches during build aren't allowed by package builders.

**Comment by burgholzer** [comment]
> pre-installed googletest is used likely due to the system path being the first.

I do not quite see how that could be happening as we do not call `find_package`, which would search for an installed version. We explicitly only ever use the vendored submodule.
That suspicion is also confirmed by my own tests on the FreeBSD VM I just set up.
A "regular build" of
```bash
git clone https://github.com/cda-tum/mqt-core --recursive
cd mqt-core
cmake -S . -B build
cmake --build build
```
does neither produce warnings nor errors (as I would expect it from our CI testing).
Maybe your definition of a "regular build" is different though and you could share whatever you use to try and build the project.
 
> Please don't use fetchcontent because fetches during build aren't allowed by package builders.

Modern versions of CMake provide an integration of `FetchContent` and `find_package` (see https://cmake.org/cmake/help/latest/guide/using-dependencies/index.html#fetchcontent-and-find-package-integration) that can also be backported to older versions of CMake.
Would you deem such a solution acceptable? This would explicitly look for an installed version of the external dependencies and only fetch them if it can't find them.
All of our external dependencies (`googletest`, `nlohmann::json`, `pybind11_json`) are available as FreeBSD ports.

**Comment by yurivict** [comment]
> I do not quite see how that could be happening as we do not call find_package [...]

This can easily happen. You don't call find_package, but the same include path is made available to the compiler through some other packages, or through a pre-set include path, and the pre-installed headers or libraries would be used instead of bundled ones. This happened before many times with other packages.

> Modern versions of CMake provide an integration of FetchContent and find_package [...]

Yes, I am aware.
You can use it, as long as pre-installed dependency packages would work through such cmake integration.

**Comment by burgholzer** [comment]
This should be fixed on `main` now by a combination of #508 and #517.
I am still working on the `FetchContent` integration in #515 and also plan to tackle the install instructions #504 before creating a new release. I'll notify you and would appreciate if you could try out the new version then.
